Adding the review doc (#1)
fayaxa/TF-K8s-Guide: TF-for-K8s-Quick-Start-Guide.md

Tungsten Fabric as a CNI for Kubernetes: A Quick Start Guide
=============================================================

> Tungsten Fabric (TF), formerly known as OpenContrail, is an open source multi-cloud, multi-stack Software Defined Networking (SDN) solution that provides a single point of control to observe, analyze and securely manage the connectivity between workloads running in the same or different domains. It essentially consists of two main components, namely: the TF controller and TF vRouter. The former collects statistics and maintains the network model and policies, whereas the latter is deployed at each host to enforce those policies. A more detailed description of the TF architecture is presented here.

Reviewer: Suggested change "...presented here." to "...covered in this document"

----

Hardware and Software Requirements
----------------------------------

Below are specified the minimum hardware and software requirements that need to be met for a smooth Tungsten Fabric experience.

Hardware:

    1. Target Nodes:
        * At least 8 vCPU cores
        * At least 32 GB RAM
        * At least 100 GB storage
       
    2. Deploy Node:
        * At least 2 CPU cores
        * At least 4 GB RAM
        * At least 50 GB storage
             
Software:

    * `CentOS 7.4` or greater
    * `Kernel release 3.10.0.862` or greater
    * `Ansible 2.6.4` or greater
    * `git`

----

Preparation
-----------

1. Generate a public ssh key in your `Deploy Node`:

>     * In command prompt of `Deploy Node` run: ssh-keygen.

Reviewer: ```suggestion
         * Make sure you are logged into your 'Deploy Node' as root (note that the deploy node is the same physical server as the k8s master node).
         * In command prompt of `Deploy Node` run: ssh-keygen.
```

    * Press <Enter> to accept all default settings.

>     * Your ssh keys would be generated and placed in the folder '.ssh' of home directory.

Reviewer: ```suggestion
    * Your ssh keys will be generated and placed in the folder '.ssh' of home directory (id_rsa and id_rsa.pub).
```

    * To view your public key, run: cat .ssh/id_rsa.pub

> 2. Copy SSH public key to the authorized key file `.ssh/authorized_keys` of `Target Nodes`. Alternatively, you can run the following command:

Reviewer: ```suggestion
2. Copy SSH public key to the authorized key file `.ssh/authorized_keys` of `Target Nodes` and 'Deploy Node. Alternatively, you can run the following command:
```

>     * ``# ssh-copy-id root@<Target_Node_IP_Address>``

Reviewer: ```suggestion
    * ``# ssh-copy-id root@<Target_Node_IP_Address>``
    * ``# ssh-copy-id root@<Deploy_Node_IP_Address>``

   
```

> 3. Install `pip` and `Ansible` in `Deploy Node`:

Reviewer: Perform package update, and install required packages on the 'Deploy Node':

>     * `` yum install -y yum-plugin-priorities yum-utils vim net-tools git ansible``

Reviewer: ```suggestion
    * `` yum update``
    * `` yum -y install git ansible ceph-common``
```

4. Clone the [TF Ansible deployer repository](https://github.com/Juniper/contrail-ansible-deployer) in `Deploy Node`:
    * `` git clone https://github.com/Juniper/contrail-ansible-deployer.git ``

----

Basic Deployment of Tungsten Fabric as a CNI for K8s
----------------------------------------------------

### Configuration

> In this section we will demonstrate a basic deployment of TF. Before proceeding we would need to configure the `instances.yml` located within the `contrail-ansible-deployer/config` folder of the cloned repository. It contains three main sections:

Reviewer: Change extension to match git clone file naming..::   instances.yml --> instances.yaml

2. `instances`: contains target host specific information and specifies which roles to install on each host.

3. `contrail_configuration`: contains TF configuration settings. All parameters in this section are optional. A list of parameters for this section is presented in `contrail_configuration.md` of the cloned repository.

The following shows a sample `instances.yaml` configuration for a k8s cluster with two nodes, one acting as a master and the other as a worker. A separate IP space for `KUBERNETES_SERVICE_SUBNETS` (service) and `KUBERNETES_POD_SUBNETS` (pods) is defined with different CIDRs. Note by default the minimum database disk space requirement is set to 256GB, however, for this deployment we have changed it to 2GB as can be seen in `DATABASE_NODEMGR__DEFAULTS__minimum_diskGB` and `CONFIG_DATABASE_NODEMGR__DEFAULTS__minimum_diskGB:`. Fill in placeholders - marked as ``<...>`` - with info from your setup:

    provider_config:
      bms:
      ssh_user: root
      ssh_public_key: <Path_to_public_key>
      ssh_private_key: <Path_to_private_key>
      domainsuffix: local
    instances:
      bms1:
       provider: bms

>        ip: <TF_contrroller_IP_address_master_node>

Reviewer: spellcheck... TF_controller_IP_address_master_node

       roles:
        config_database:
        config:
        control:
        analytics:
        analytics_database:
        webui:
        k8s_master:
        kubemanager:

>         k8s_node:

Reviewer: DELETE LINE

>         #vrouter:

Reviewer: DELETE LINE

      bms2:
       provider: bms
       ip: <Worker_IP_address>
       roles:
        vrouter:
        k8s_node:
    global_configuration:

>       CONTAINER_REGISTRY: opencontrailnightly

Reviewer: ```suggestion
      CONTAINER_REGISTRY: tungstenfabric
```

    contrail_configuration:

>       CONTAINER_REGISTRY: opencontrailnightly

Reviewer: DELETE LINE

      CONTRAIL_VERSION: latest
      KUBERNETES_CLUSTER_PROJECT: {}
      KUBERNETES_IP_FABRIC_SNAT: True

>       KUBERNETES_POD_SUBNETS: 10.64.0.0/12

Reviewer: DELETE LINE

>       KUBERNETES_SERVICE_SUBNETS: 10.128.0.0/12

Reviewer: DELETE LINE

      DATABASE_NODEMGR__DEFAULTS__minimum_diskGB: 2
      CONFIG_DATABASE_NODEMGR__DEFAULTS__minimum_diskGB: 2

>       #KUBERNETES_PUBLIC_FIP_POOL: {'domain': 'default-domain', 'project': 'k8s-default', 'network': 'external', 'name': 'default'}

Reviewer: DELETE LINE

    == Contrail control ==
    control: active
    nodemgr: active
    named: active
    dns: active

    == Contrail config-database ==
    nodemgr: active
    zookeeper: active
    rabbitmq: active
    cassandra: active

    == Contrail kubernetes ==
    kube-manager: active

    == Contrail database ==
    nodemgr: active
    query-engine: active
    cassandra: active

    == Contrail analytics ==
    api: active
    nodemgr: active
    collector: active

    == Contrail webui ==
    web: active
    job: active

    == Contrail vrouter ==
    nodemgr: active
    agent: active

    == Contrail device-manager ==

    == Contrail config ==
    svc-monitor: active
    nodemgr: active
    device-manager: active
    api: active
    schema: active

Alternatively, the installation can also be verified by accessing the TF web interface via port `8143` over `HTTPS`:

    https://<TF_contrroller_IP_address_master_node>:8143
    Default credentials: 
        * username: admin
        * password: contrail123

![](ui.png)

----

Use Cases
---------

> So far we have configured and deployed the TF framework on to as a CNI for a Kubernetes cluster. The rest of this guide will walk you through a few common scenarios around networking that you can encounter when developing and operating an application that runs on Kubernetes. For the purpose of demonstration, we will be using a very simple, light-weight and well documented mock application called `yelb`. Further details on its working and internal architecture can be found [here](https://github.com/mreferre/yelb#yelb-architecture). It essentially comprises of four components: `yelb-ui` the frontend UI, `yelb-appserver`, `yelb-db` postgres backend database and `yelb-cache` redis cache server.

Reviewer: Replace first sentence with...
 So far we have configured and deployed the TF framework as a CNI for a Kubernetes cluster.

> So far we have configured and deployed the TF framework on to as a CNI for a Kubernetes cluster. The rest of this guide will walk you through a few common scenarios around networking that you can encounter when developing and operating an application that runs on Kubernetes. For the purpose of demonstration, we will be using a very simple, light-weight and well documented mock application called `yelb`. Further details on its working and internal architecture can be found [here](https://github.com/mreferre/yelb#yelb-architecture). It essentially comprises of four components: `yelb-ui` the frontend UI, `yelb-appserver`, `yelb-db` postgres backend database and `yelb-cache` redis cache server.

Reviewer: Replace second sentence with...
The rest of this guide will walk you through a few common scenarios around networking that may be encountered when developing and operating an application that runs on Kubernetes.

> So far we have configured and deployed the TF framework on to as a CNI for a Kubernetes cluster. The rest of this guide will walk you through a few common scenarios around networking that you can encounter when developing and operating an application that runs on Kubernetes. For the purpose of demonstration, we will be using a very simple, light-weight and well documented mock application called `yelb`. Further details on its working and internal architecture can be found [here](https://github.com/mreferre/yelb#yelb-architecture). It essentially comprises of four components: `yelb-ui` the frontend UI, `yelb-appserver`, `yelb-db` postgres backend database and `yelb-cache` redis cache server.

Reviewer: Replace last sentence with...
It is comprised of four components

>     #Switch to Target Node 1 (Master Node) and make sure you have

Reviewer: ```suggestion
    #Switch to  the Master Node and make sure you are root
```

>     #switched to the root user before proceeding further:

Reviewer: ```suggestion
    #before proceeding further:
```

    sudo -i
    whoami | grep root || sudo -s

>     #Install git:

Reviewer: DELETE LINE

>     yum -y install git

Reviewer: DELETE LINE
